### PR TITLE
TerminateAll on local:docker

### PR DESCRIPTION
**Pull request by coryschwartz**
Make terminate work on local docker

My state before:
```
↬ docker ps
CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS              PORTS               NAMES
6db2df072860        ipfs/testground:latest   "testground sidecar …"   3 minutes ago       Up 3 minutes                            testground-sidecar
f8329deca706        redis                    "redis-server"           3 minutes ago       Up 3 minutes                            testground-redis
467421a30913        goproxy/goproxy          "/goproxy"               6 minutes ago       Up 5 minutes        8081/tcp            testground-goproxy
```

Run the terminate command:
```
↬ ./testground terminate --runner local:docker
Feb 20 16:49:35.575065	INFO	testground client initialized	{"addr": "localhost:8042"}
terminating all jobs on runner local:docker
all jobs on runner local:docker were terminated

↬ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

```

Logs from the server:
```
↬ ./testground daemon
Feb 20 16:49:30.664818	INFO	listen and serve	{"addr": "127.0.0.1:8042"}
Feb 20 16:49:30.665043	INFO	daemon listening	{"addr": "127.0.0.1:8042"}
Feb 20 16:49:35.578276	INFO	Terminate local:docker requested
Feb 20 16:49:35.587458	INFO	deleting containers	{"ids": ["6db2df072860a89032caff9d9d7d5447aed3a8f614989ba624d56d1733e6e217", "f8329deca70683b32aa86218333e57a220f90eceae571e7f92f4bd22c4917177", "467421a30913122600f10ea9094e4b45c3eeab31ad667448f25b6f8c728bd015"]}
Feb 20 16:49:37.480049	INFO	Deleted testground containers.
Feb 20 16:49:37.480087	INFO	To delete networks and images, you may want to run `docker system prune`
```

License: MIT
Signed-off-by: Cory Schwartz <cory@protocol.ai>

### CODE REVIEW

**Comment by coryschwartz** [comment]
I added killing the long-running test plan containers also.

It works by adding a label during the run phase, and filtering them out during the terminate phase.

File: pkg/runner/local_docker.go [hunk 1]
```diff
@@ -474,3 +475,61 @@ func (*LocalDockerRunner) ConfigType() reflect.Type {
...
+	// It seems like you have to do two separate queries for these filter types. If this isn't the
+	// case, this could be made simpler. Do two queries to the local docker daemon. For one, filter
+	// out the names of specific container infrastructure.
+	// For the other filter out where labels match the testground.purpose, which we apply to all
+	// plan containers managed by testground
+	listOpts := types.ContainerListOptions{}
+	listOpts.Filters = filters.NewArgs()
+	listOpts.Filters.Add("name", "testground-sidecar")
+	listOpts.Filters.Add("name", "testground-redis")
+	listOpts.Filters.Add("name", "testground-goproxy")
```
Comment (raulk) [inline pkg/runner/local_docker.go]: The proxy is owned by the `docker:go` builder. We could make builders implement the `Terminator` interface too if needed, but I wouldn't delete it here.

File: pkg/runner/local_docker.go [hunk 2]
```diff
@@ -474,3 +475,61 @@ func (*LocalDockerRunner) ConfigType() reflect.Type {
...
+	if err != nil {
+		return err
+	}
+	plancontainers, err := cli.ContainerList(ctx, planListOpts)
+	if err != nil {
+		return err
+	}
+
+	var containerIds []string
+	if err != nil {
```
Comment (raulk) [inline pkg/runner/local_docker.go]: Out of place err check.

File: pkg/runner/local_docker.go [hunk 3]
```diff
@@ -474,3 +475,61 @@ func (*LocalDockerRunner) ConfigType() reflect.Type {
...
+		return err
+	}
+	for _, container := range infracontainers {
```
Comment (raulk) [inline pkg/runner/local_docker.go]: Can use an append here to simplify this:

```go
containers := append(infracontainers, plancontainers...)
```

File: pkg/runner/local_docker.go [hunk 4]
```diff
@@ -474,3 +475,61 @@ func (*LocalDockerRunner) ConfigType() reflect.Type {
...
+func (*LocalDockerRunner) TerminateAll() error {
+	log := logging.S()
+	log.Info("Terminate local:docker requested")
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+	// Build query for testground infrastructure
```
Comment (raulk) [inline pkg/runner/local_docker.go]: nit: Vertical spacing would be welcome ;-)

File: pkg/runner/local_docker.go [hunk 5]
```diff
@@ -474,3 +475,61 @@ func (*LocalDockerRunner) ConfigType() reflect.Type {
...
+	// It seems like you have to do two separate queries for these filter types. If this isn't the
+	// case, this could be made simpler. Do two queries to the local docker daemon. For one, filter
+	// out the names of specific container infrastructure.
+	// For the other filter out where labels match the testground.purpose, which we apply to all
+	// plan containers managed by testground
```
Comment (raulk) [inline pkg/runner/local_docker.go]: Thanks for the comments ;-)

File: pkg/runner/local_docker.go [hunk 6]
```diff
@@ -474,3 +475,61 @@ func (*LocalDockerRunner) ConfigType() reflect.Type {
...
+	}
+	for _, container := range plancontainers {
+		containerIds = append(containerIds, container.ID)
+	}
+
+	err = deleteContainers(cli, log, containerIds)
+	if err != nil {
+		log.Info("Failed to delete testground containers")
+	}
+	log.Info("To delete networks and images, you may want to run `docker system prune`")
```
Comment (raulk) [inline pkg/runner/local_docker.go]: nit: log lines start with lower-case.

File: pkg/runner/local_docker.go (hunk 3 above)
Comment (raulk) [reply pkg/runner/local_docker.go]: Ah, realised that you're extracting the ID field.

File: pkg/runner/local_docker.go [hunk 7]
```diff
@@ -474,3 +475,61 @@ func (*LocalDockerRunner) ConfigType() reflect.Type {
...
+	for _, container := range infracontainers {
+		containerIds = append(containerIds, container.ID)
+	}
+	for _, container := range plancontainers {
+		containerIds = append(containerIds, container.ID)
+	}
+
+	err = deleteContainers(cli, log, containerIds)
+	if err != nil {
+		log.Info("Failed to delete testground containers")
```
Comment (raulk) [inline pkg/runner/local_docker.go]: We should be wrapping this error and returning it?

File: pkg/runner/local_docker.go [hunk 8]
```diff
@@ -474,3 +475,61 @@ func (*LocalDockerRunner) ConfigType() reflect.Type {
...
+	planListOpts.Filters = filters.NewArgs()
+	planListOpts.Filters.Add("label", "testground.purpose=plan")
+
+	infracontainers, err := cli.ContainerList(ctx, listOpts)
+	if err != nil {
+		return err
+	}
+	plancontainers, err := cli.ContainerList(ctx, planListOpts)
+	if err != nil {
+		return err
```
Comment (raulk) [inline pkg/runner/local_docker.go]: Good practice to wrap errors with more context:

```go
return fmt.Errorf("failed to list test plan containers: %w", err)
```

File: pkg/runner/local_docker.go [hunk 9]
```diff
@@ -474,3 +475,61 @@ func (*LocalDockerRunner) ConfigType() reflect.Type {
...
 	return []string{"docker:go"}
 }
+
+// This method deletes the testground containers.
+// It does *not* delete any downloaded images or networks.
+// I'll leave a friendly message for how to do a more complete cleanup.
+func (*LocalDockerRunner) TerminateAll() error {
+	log := logging.S()
+	log.Info("Terminate local:docker requested")
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (raulk) [inline pkg/runner/local_docker.go]: We should be using the request from the context, I'll wire it in.

**Pull request merged**